File: Sample-code/WILLDevicesSample-ObjC/WILLDevicesSample-ObjC/FileTransferViewController.m
//
//  FileTransferViewController.m
//  WILLDevicesSample-ObjC
//
//  Created by Joss Giffard-Burley on 08/12/2017.
//  Copyright © 2017 Wacom. All rights reserved.
//

#import "FileTransferViewController.h"
#import "AppDelegate.h"

@import WILLDevicesCore;

@interface FileTransferViewController ()

/// The list of downloaded documents from the device
@property (nonatomic, strong) NSMutableArray<InkDocument *> *downloadedDocuments;

/// Background download queue
@property (nonatomic, strong) dispatch_queue_t downloadQueue;

/// The collection view used to render samples of the files recevied
@property (nonatomic, strong) IBOutlet UICollectionView *collectionView;

/// Flag to stop spamming when we are polling for files
@property (nonatomic, assign) BOOL showFinishedPrompt;

/// Flag to see if we should be polling for new files
@property (nonatomic, assign) BOOL pollForNewFiles;

/// Inkdevice service for file download
@property (nonatomic, strong) id<FileTranserService> fileService;

@end

@implementation FileTransferViewController

//========================================================================================================
// MARK: UIView Methods
//========================================================================================================

- (void)viewDidLoad {
    [super viewDidLoad];
    self.downloadedDocuments = [NSMutableArray new];
    self.downloadQueue = dispatch_queue_create("download", NULL);
    self.showFinishedPrompt = YES;
    self.pollForNewFiles = YES;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if(self.inkDevice == NULL || self.inkDevice.deviceStatus == DeviceStatusNotConnected) {
        [AppDelegate postNotification:@"Error" withBody:@"InkDevice not connected"];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            self.fileService = NULL;
            [self.navigationController popToRootViewControllerAnimated:YES];
        });
    }
    //Attempt to start the ink service
    @try {
        self.fileService = (id<FileTranserService>)[self.inkDevice getService:InkDeviceServiceTypeFileTransfer error:NULL];
    } @catch (NSException *e) {
        NSString *errString = [NSString stringWithFormat:@"Failed to start file transferservice:%@", e.reason];
        [AppDelegate postNotification:@"Error" withBody:errString];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            self.fileService = NULL;
            [self.navigationController popToRootViewControllerAnimated:YES];
        });
    }
}

- (void)viewDidAppear:(BOOL)animated {
    CGFloat tileSize = 288.0f;
    self.pollForNewFiles = YES;
    
    @try {
    dispatch_async(self.downloadQueue, ^{
        self.fileService.dataReceiver = self;
        
        CGFloat xScale = tileSize / self.deviceHeight;
        CGFloat yScale = tileSize / self.deviceWidth;
        CGFloat scale = fmin(xScale, yScale);
        self.fileService.transform = CGAffineTransformIdentity;
        CGAffineTransform t = CGAffineTransformIdentity;

        if(self.shouldRotateImages) {
            CGFloat rotationAngle = (-M_PI/2.0) + M_PI;
            t = CGAffineTransformScale(t, scale, scale);
            t = CGAffineTransformRotate(t, rotationAngle);
            t = CGAffineTransformConcat(t, CGAffineTransformMake(1, 0, 0, 1, tileSize, 0));
        } else {
            t = CGAffineTransformScale(t, scale, scale);
            [self.fileService setTransform:CGAffineTransformScale(CGAffineTransformIdentity, scale, scale)];
        }
        [self.fileService setTransform:t];

        [self.fileService startWithProvideRawData:YES error:NULL completionHandler:NULL];
    });
    } @catch (NSException *e){
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [AppDelegate postNotification:@"Error" withBody:[NSString stringWithFormat:@"%@:%@", @"Failed to start file transder", e.reason]];
            [self.navigationController popToRootViewControllerAnimated:YES];
        });
    }
}

- (void)viewWillDisappear:(BOOL)animated {
    self.pollForNewFiles = NO;
    NSError *e;
    [self.fileService endAndReturnError:&e completionHandler:NULL];
    
    if(e != NULL) {
        [AppDelegate postNotification:@"Error" withBody:[NSString stringWithFormat:@"%@: %@", @"Error stopping file transfer service", e.localizedDescription]];
    }
    
}
//========================================================================================================
// MARK: File data delegate
//========================================================================================================

- (void)errorWhileDownloadingFile:(NSError * _Nonnull)error {
    NSLog(@"%@", [NSString stringWithFormat:@"Error during download: %@", error.localizedDescription]);
}

- (void)noMoreFiles {
    if(self.showFinishedPrompt) {
        [AppDelegate postNotification:@"Complete" withBody:@"No more files to download from smartpad"];
    }
    
    self.showFinishedPrompt = NO;
    [self.fileService endAndReturnError:NULL completionHandler:NULL];
    
    //Poll for new files
    __weak FileTransferViewController *w = self;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        if([w pollForNewFiles]) {
            [w.fileService startWithProvideRawData:YES error:NULL completionHandler:NULL];
        }
    });
}



- (enum FileDataReceiverStatus)receiveFileWithFileData:(InkDocument *)fileData remainingFilesCount:(NSInteger)remainingFilesCount {
    [self.downloadedDocuments addObject:fileData];
    dispatch_async(dispatch_get_main_queue(), ^{
        self.showFinishedPrompt = YES;
        NSLog(@"New file recevied from device");
        [self.collectionView reloadData];
    });
    
    //We saved the file that was returned by the device. Sending 'FILESAVED' back to the device will cause the file to be removed from the device memory
    return FileDataReceiverStatusFileSaved;
}

//========================================================================================================
// MARK: UICollection view delegates / datasource
//========================================================================================================

- (nonnull __kindof UICollectionViewCell *)collectionView:(nonnull UICollectionView *)collectionView cellForItemAtIndexPath:(nonnull NSIndexPath *)indexPath {
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"previewCell" forIndexPath:indexPath];

    InkDocument *document = self.downloadedDocuments[indexPath.row];
    
    //Get a bezier curve for the document. As we set a scale when the document was downloaded, we can just use the
    //property from the InkDocument directly
    InkGroup *root = [document getRoot];

    void (^nodeProcessor)(InkNode *node);
    
    //Block to iterate the nodes
    nodeProcessor = ^(InkNode *node){
        if([node isKindOfClass:[InkGroup class]]) {
            InkGroup *groupNode = (InkGroup *)node;
            NSInteger currentIdx = 0;
            InkNode *child = [groupNode getChildWithIndex:currentIdx++];
            while (child != NULL) { //Iterate thrould child nodes
                if([child isKindOfClass:[InkStroke class]]) {
                    InkStroke *stroke = (InkStroke *)child;
                    UIBezierPath *bezPath = [stroke bezierPath];
                    if(bezPath != NULL) {
                        CAShapeLayer *shapeLayer = [[CAShapeLayer alloc] init];
                        shapeLayer.path = bezPath.CGPath;
                        shapeLayer.position = CGPointZero;
                        shapeLayer.fillColor = [UIColor blackColor].CGColor;
                        shapeLayer.strokeColor = [UIColor clearColor].CGColor;
                        [cell.layer addSublayer:shapeLayer];
                    }
                }
                
                if([child isKindOfClass:[InkGroup class]]) {
                    nodeProcessor(child);
                }
                child = [groupNode getChildWithIndex:currentIdx++];
            }
        }
    };
    
    nodeProcessor(root);
    return(cell);
}

- (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return([self.downloadedDocuments count]);
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return(1);
}


@end

File: Sample-code/WILLDevicesSample/WILLDevicesSample/FileTransferViewController.swift
//
//  FileTransferViewController.swift
//  WILLDevicesSample
//
//  Created by Joss Giffard-Burley on 17/07/2017.
//  Copyright © 2017 Wacom. All rights reserved.
//

import UIKit
import WILLDevices
import WILLDevicesCore
import WILLInk


/// This view visualises files downloaded from the device. Once a file has been downloaded from the device, it
/// will be removed. Files are displayed in a simple collection view that displays the UIBezierCurve of the strokes
/// for the file as the file thumbnail.
class FileTransferViewController: UIViewController {
    //========================================================================================================
    // MARK: Properties
    //========================================================================================================
    
    /// The connected ink device
    weak var inkDevice: InkDevice?
    
    /// The file transfer service provided by the device
    var fileService: FileTranserService?
    
    public var deviceWidth: CGFloat = 100.0
    public var deviceHeight: CGFloat = 100.0
    
    /// The list of downloaded documents from the device
    var downloadedDocuments = [InkDocument]()
    
    /// Background download queue
    let downloadQueue = DispatchQueue(label: "download")
    
    /// The collection view used to render samples of the files recevied
    @IBOutlet var collectionView: UICollectionView!
    
    /// Flag to stop spamming when we are polling for files
    internal var showFinishedPrompt = true
    
    /// Flag to see if we should be polling for new files
    internal var pollForNewFiles = true
    
    /// SHould we rotate the recevied files to match orientation
    internal var shouldRotateImages = true
    
    //========================================================================================================
    // MARK: UIView Methods
    //========================================================================================================
    
    override func viewWillAppear(_ animated: Bool) {
        if inkDevice == nil || (inkDevice?.deviceStatus ?? .notConnected) == .notConnected {
            AppDelegate.postNotification("Error", bodyText: "InkDevice not connected")
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                self.fileService = nil
                self.navigationController?.popViewController(animated: true)
            }
        }
        //Attempt to start the ink service
        do {
            try self.fileService = self.inkDevice?.getService(.fileTransfer) as? FileTranserService
        } catch (let e) {
            AppDelegate.postNotification("Error", bodyText: "Failed to start file transferservice:\(e.localizedDescription)")
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { 
                self.fileService = nil
                self.navigationController?.popViewController(animated: true)
            }
        }
        
    }
    
    override func viewDidAppear(_ animated: Bool) {
        let tileSize = CGFloat(288.0)
        pollForNewFiles = true

        downloadQueue.async {
        do {
            self.fileService?.dataReceiver = self //Receive the point data
            //Set the scale to match the render tile size.
            let xScale = tileSize / self.deviceHeight
            let yScale = tileSize / self.deviceWidth
            let scale = fmin(xScale, yScale) //So we have 1:1 scale
            
            if self.shouldRotateImages {
                let rotationAngle = (-CGFloat.pi/2.0) + .pi
                self.fileService?.transform = CGAffineTransform(scaleX: scale, y: scale).rotated(by: rotationAngle).concatenating(CGAffineTransform(a: 1, b: 0, c: 0, d: 1, tx: tileSize, ty: 0))
            } else {
                self.fileService?.transform = CGAffineTransform(scaleX: scale, y: scale)
            }
            
            try self.fileService?.start(provideRawData: true) {success, error in
                print("file service start")
                print("success -> \(success)")
                print("error -> \(String(describing: error))")
            }
        } catch (let e) {
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                AppDelegate.postNotification("Error", bodyText: "Failed to start realtime ink service:\(e.localizedDescription)")
                self.navigationController?.popViewController(animated: true)
            }
            }
        }
        AppDelegate.postNotification("Starting file download", bodyText: "Getting files from device")
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        //Stop service and dispose
        guard let s = fileService else {
            return
        }
        pollForNewFiles = false
        do {
            try s.end(completionHandler: { (success, error) in
                print("file service end")
                print("success -> \(success)")
                print("error -> \(String(describing: error))")
            })
        } catch (let e) {
            AppDelegate.postNotification("Error", bodyText: "Error closing ink serivce:\(e.localizedDescription)")
        }
    }
    
}

//========================================================================================================
// MARK: File data delegate
//========================================================================================================

extension FileTransferViewController: FileDataReceiver {
    func noMoreFiles() {
        if showFinishedPrompt {
            AppDelegate.postNotification("Complete", bodyText: "No more files to download from smartpad")
        }
        
        showFinishedPrompt = false
        ((try? fileService?.end(completionHandler: { (success, error) in
            print("file service end")
            print("success -> \(success)")
            print("error -> \(String(describing: error))")
        })) as ()??)
        
        //Poll for new files
        downloadQueue.asyncAfter(deadline: .now() + 3.0) { [weak self] in
            guard let s = self?.fileService else {
                return
            }
            
            //If the service isn't started, start it
            if (self?.pollForNewFiles ?? false) {
                try? s.start(provideRawData: true, completionHandler: { (success, error) in
                    print("file service start")
                    print("success -> \(success)")
                    print("error -> \(String(describing: error))")
                })
            }
        }
    }
    
    func receiveFile(fileData: InkDocument, remainingFilesCount: Int) -> FileDataReceiverStatus {
        //Add the document to our list and update the collection view
        downloadedDocuments.append(fileData)
        DispatchQueue.main.async {
            self.showFinishedPrompt = true
            print("New file received from device")
            self.collectionView.reloadData()
        }
        
        //We return file saved. This causes the file to be deleted from the device
        return .fileSaved
    }
    
    func errorWhileDownloadingFile(_ error: Error) {
        print("Error during download:\(error.localizedDescription)")
    }
}

//========================================================================================================
// MARK: UICollection view delegates / datasource
//========================================================================================================


extension FileTransferViewController: UICollectionViewDelegate, UICollectionViewDataSource {
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return downloadedDocuments.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "previewCell", for: indexPath)
        
        if let sublayers = cell.layer.sublayers {
            for l in sublayers {
                if l is CAShapeLayer {
                    l.removeFromSuperlayer()
                }
            }
        }
        
        let document = downloadedDocuments[indexPath.row]
        
        //Get a bezier curve for the document. As we set a scale when the document was downloaded, we can just use the
        //property from the InkDocument directly
        var documentIterator = document.getRoot().makeIterator()
        
        while documentIterator.hasNext() {
            guard let node = documentIterator.next() else {
                print("Error getting next node on document iterator!")
                break
            }
            
            if node is InkStroke, let bezPath = (node as! InkStroke).bezierPath {
                let shapeLayer = CAShapeLayer()
          
                shapeLayer.path = bezPath.cgPath
                shapeLayer.position = CGPoint.zero
                shapeLayer.fillColor = UIColor.black.cgColor
                shapeLayer.strokeColor = UIColor.clear.cgColor
                cell.layer.addSublayer(shapeLayer)
            }
        }
 
        return cell
    }
}
